[ML] Transforms/Data frame analytics: Aligns data view / destination index creation workflow in wizards

walterra opened the pull request:
## Summary

Fixes #160593.

Consolidates UI elements and backend code to create/delete data views and destination indices related to transforms and data frame analytics.

We ended up with two different approaches for creating data views in the wizards for transforms and data frame analytics, the original reason was we were not aware of the `allowNoIndex: true` setting and worked around that in different ways. 

This PR aligns UI workflows and moves related code to a new package `@kbn/ml-data-view-utils` for data views and `@kbn/ml-creation-wizard-utils` for the destination index form. The latter might be used for other shared components across wizard..

In Data Frame Analytics, the checkbox to create a data view was removed from the last "Create" step, instead the option to create a data view was moved to the "Details" step. 

In Transforms, the UI component to create the destination index was brought over from DFA where there is a switch option to automatically use the job ID as the name for the destination index by default.

Updated data frame analytics wizard:

<img width="1120" alt="image" src="https://github.com/elastic/kibana/assets/230104/fe8e6db2-ced5-48c0-9cfb-3ab3076560b6">

<img width="1101" alt="image" src="https://github.com/elastic/kibana/assets/230104/7a00b7a1-8030-4e65-a52d-34f6380fcddd">

Updated transform wizard:

<img width="1050" alt="image" src="https://github.com/elastic/kibana/assets/230104/b2aba800-ebfe-429e-b01c-95557506805f">



### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

walterra commented:
I accidentally switched this one to review but it's not finished yet, reverting back to draft and will ping reviewers later again.

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

peteharverson commented on x-pack/packages/ml/data_view_utils/components/create_data_view_switch.tsx:
```diff
@@ -0,0 +1,35 @@
+  createDataView,
+  setCreateDataView,
+}) => (
+  <EuiSwitch
```
You need to add the 'Create data view' switch into the end of the advanced JSON editor for DFA jobs - this is how it looks on `main` currently:

<img width="526" alt="image" src="https://github.com/elastic/kibana/assets/7405507/07e2266b-0582-4a03-95fa-b4f0396dd158">

peteharverson reviewed (commented):
Tested and overall looks good. Just need to add the 'Create data view' switch into the DFA advanced editor.

walterra commented:
@peteharverson Thanks for catching the missing data view section in the JSON editor. Added the section to create data views back to the JSON editor variant. Also updated the toggle to auto-start the DFA job from checkbox to switch to be in line with the updated data view UI.

<img width="1391" alt="image" src="https://github.com/elastic/kibana/assets/230104/68f5cd6e-a653-4489-86fd-1a2bf8b85f89">

peteharverson commented on x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_creation/components/create_step/create_step.tsx:
```diff
@@ -75,125 +64,63 @@ export const CreateStep: FC<Props> = ({ actions, state, step }) => {
+                <EuiSwitch
+                  data-test-subj="mlAnalyticsCreateJobWizardStartJobSwitch"
+                  name="mlAnalyticsCreateJobWizardStartJobSwitch"
+                  disabled={canCreateDataView === false}
```
Why is the enabled state of this switch linked to `canCreateDataView`?

walterra replied:
Good find, thanks! That was a copy/paste mistake, the original checkbox didn't have a `disabled` attribute so I removed it again from the switch in [0b8692e](https://github.com/elastic/kibana/pull/171202/commits/0b8692ea132db91593f7efe7ef55272273b469c0).

alvarezmelissa87 reviewed (approved):
LGTM ⚡
